### ISSUE TITLE
Allow entity reference fields to be optional

### ISSUE THREAD

**Issue by jonathanjfshaw** [jhedstrom/drupalextension#346]
There are some wrinkles in declaring entities with optional entity reference fields.
For example,
(A)
```
    Given "article" content:
      | title       | field_reference    | 
      | Test1     |                            |
      | Test2     | Test1                   |
```
fails with "No entity '' of type 'node' exists." because in the first row the field_reference column is blank, which gets passed to the driver as an empty string, which doesn't match the label of any node. Things get weirder when you try to reference users and leave a row blank, because in that case a reference is created to user 0 (the anonymous user).

It is possible to workaround this by creating 2 givens:
(B)
```
    Given "article" content:
      | title       |
      | Test1     |
     Given "article" content:
      | title       | field_reference    | 
      | Test2     | Test1                   |
```
but this is inelegant.

I think it would be ideal if non-required entity reference fields could be blank as in (A), in which case no reference is created. And certainly it would be less confusing if user references behaved the same as node ones; if people want to reference the anonymous user, perhaps they should have to do that explicitly.

Implementation of this is probably a driver matter; I've started discussion here in the extension because it starts with a question of how we want step tables to be understood.

**Comment by jonathanjfshaw** [comment]
Filtering out zero-length string columns in parseEntityFields in rawDrupalContext seems to work OK:
```
        // Replace regular fields inline in the entity after parsing.
        if (!$is_multicolumn) {
          $entity->$field_name = $values;
          if ($field_value === '') {
            unset($entity->$field);
          }
        }
      }
    }

    // Add the multicolumn fields to the entity.
    foreach ($multicolumn_fields as $field_name => $columns) {
      if (!empty(array_filter($columns, function ($var) {return ($var === '');}))) {
        $entity->$field_name = $columns;
      }
    }
  }
```

**Comment by jonathanjfshaw** [comment]
Thinking about this more, my suggested solution of filtering out zero-length strings in parseEntityFields is no good. A zero-length string is a meaningful value for a text field, when the field has a default value that is not a zero-length string, so this can't be applied to all fields. 

ParseEntityFields doesn't know about field types; logic that is field type specific needs to be in the field handlers.

I don't think there's any simple value that a field handler could return to indicate a field should be unset, because all possible return values (including e.g. NULL) are also legitimate default-overriding values for some field type or other.

Perhaps the solution would be to add a new public property to the AbstractHandler `public $unsetIfEmpty = false;` which could be overriden to be true on the EntityReferenceHandler. In parseEntityFields we could then interrogate this property and unset the field conditionally.

**Comment by pfrenssen** [comment]
> Thinking about this more, my suggested solution of filtering out zero-length strings in parseEntityFields is no good. A zero-length string is a meaningful value for a text field, when the field has a default value that is not a zero-length string, so this can't be applied to all fields.

I think we can keep it simpler by assuming that an empty field simply means that nothing is being filled in. So if a default value would already be present in the field then this will be retained.

If an existing default value needs to be forcefully cleared in the course of a scenario this can better be explicitly mentioned in a step. This will be much clearer to the non-technical audience for which the scenarios are intended.

**Comment by jonathanjfshaw** [comment]
Agreed, implemented in PR #349.
